Share libs between node and deltachat

flub opened the pull request:
The most important change is to build node.js with libssl and libz as
shared libraries.  Later on deltachat-core gets build using these same
libraries and all the objects can happily live together in the same
process.  Before we relied a bit on luck that the versions were
compatible.

Secondly this adds a readme, this is a work in progress.

Than the smaller changes:

- Clean up unused files in libsasl to reduce the result a little.

- Be more explicit about the build system used rather than let it
  autodetect.

- Pass options to the libsasl build as well, to customise the build a
  bit better.

- Remove a few unneeded chromedriver sources

- Document the required environment variables a little.

- Remove redundant `npm run build` command as this is covered in the
  `npm run pack` one (in packages.js).

muelli commented on README.md:
````diff
@@ -0,0 +1,73 @@
+python3 ../flatpak-builder-tools/npm/flatpak-npm-generator.py package-lock.json
+```
+
+This will produce the `generaed-sources.json` file which is referenced
````
typos here.
`generated-sources.json`.  The "t" is missing.
The line below has a typo in  `desktop`.

flub replied:
thanks, done

flub commented:
ping?

muelli commented:
any reason for removing the ARM downloads? It doesn't work right now (I forgot why) but my hope is that having the ARM downloads performed encourages someone to work on it and make it work.

muelli commented:
FWIW: The way I generate the `generated-sources.json` is to checkout the latest deltachat release, i.e. `git checkout -b 0.101.0 v0.101.0`, there, I have a Makefile with the following content:

```
generated-sources.json: package-lock.json
	~/vcs/flatpak-builder-tools/npm/flatpak-npm-generator.py --recursive package-lock.json

package-lock.json: package.json
	npm install --package-lock-only

```

so I type make.

Then I copy the resulting files to the flatpak directory, i.e. `cp package-lock.json generated-sources.json ../deltachat-desktop.flatpak/`.

muelli commented:
I've taken most of your changes on the manifest. There were no changes on the definition of the delta app except for removing comments which I deem helpful enough, like that certain files ought to be taken from the build rather than us shipping them, like the appstream or icon files. Let me know if I've missed anything.

flub commented on README.md:
```diff
@@ -0,0 +1,73 @@
+### Uploading to flathub
+
+- signing?
+- upload
```
@muelli anything which needs to be added to this section?  As I understand currently any commit here is automatically build and signed by flathub so we don't have to do any management of this?

Should we be signing commit to this repo?  http://docs.flatpak.org/en/latest/flatpak-builder.html#signing seems to suggest so but this seems not enforced?  Also not sure if the way they describe there works for this situation?

muelli replied:
that signing is part of ostree commits, not git. I see how this can cause confusion and it could do with a more explicit mentioning. OTOH, "git" is nowhere mentioned at all on that page so it might be fair to not assume git being dealt with in first place.

And your assessment is correct, every push to master (or beta, for completeness sake) causes a build with downloadable artefacts. So no management needed.

flub replied:
Oh, I see how I read that wrong.  Thanks for explaining.  Maybe I just mention that every commit here results in a new published release.

flub commented:
Hey, I'm not really a fan of how you've created some commits in my name which I never made, it confuses the history somewhat.  I'd rather chat about the proposed changed and incorporate feedback here until we're both happy with the result.  Even if you want to commit some stuff already, using your own commits is fine - if you really think attribution matters that much, mentioning it in the commit message will do (but I don't mind).  Anyway, I've managed to resolve all resulting conflicts and updated the PR with the remainder of things I think are still useful and updated the commit message to reflect so.

With respect to the ARM downloads, they are unused so result in unneeded network traffic and storage etc.  I don't think not having them would be a deterrent to anyone wanting to port to arm.  Hence why I'd remove them.  If we really want to keep them then maybe it makes sense to start an arm branch and have them there?

For the flatpak-npm-generator.py call since there is only one lock file in the deltachat-desktop repo and it's in the root there's no need to add --recursive to this command I think.  I think the current description in the readme matches the steps you describe you take?  It would certainly be nice to streamline and automate this in the future.  For the moment we have https://github.com/deltachat/deltachat-desktop/issues/694 which would be a first step into this direction.

Lastly, as I indicated on the updated commit message, I currently think it's the flatpak's responsibility to provide the desktop integration files and not the deltachat-dekstop, hence why I changed the comments that way.  My goal is to remove the electron-builder and AppImage etc stuff from deltachat-desktop once we're comfortable with releasing the flatpak as part of normal deltachat releases.  So we can't rely on those steps.

Oh, and I still think the readme adds value for  people new to the repository btw.  It's essentially a bunch of notes to myself that I'd have liked to have.

muelli commented:
> Hey, I'm not really a fan of how you've created some commits in my name which I never made

oh, I'm sorry about that. I didn't mean to offend. At all. In fact, quite the opposite. Should I force push a new history?

> With respect to the ARM downloads, they are unused so result in unneeded network traffic and storage etc

yeah, it'll save some bandwidth. The resulting flatpak won't be affected, though. I'm fine with not carrying cruft along, but still, building for x86-64 only seems unnecessary.  But you're proposal to carry the ARM stuff around in a new branch is good.

> For the flatpak-npm-generator.py call since there is only one lock file in the deltachat-desktop repo and it's in the root there's no need to add --recursive to this command I think.

I got bitten by this once. That is, I needed to find out about the recursive call to make it work.  That said, not having to make that call because upstream ships that file would certainly be appreciated.




> Lastly, as I indicated on the updated commit message, I currently think it's the flatpak's responsibility to provide the desktop integration files and not the deltachat-dekstop

ah, this is unusual, because there are probably other downstreams which care about the `.desktop` file, the `.appdata` file, the icons, etc., too. AFAIU flatpaks try to be as lean as possible, shipping as few custom patches or files as possible. And spontaneously, I'd say that the upstream of a desktop app is the very correct place to handle those.  And it currently does, right?  If you build the `.deb` file then you get the icon. I would expect a .`desktop` file to be generated, too. It doesn't seem smart to make every downstream ship the same file themselves.

muelli commented on README.md:
````diff
@@ -0,0 +1,73 @@
+`base-version` properties for the current version numbers.
+
+```
+flatpak install flathub \
````
given the level of detail, it might be worth linking towards https://flatpak.org/setup/ and maybe even copy and pasting the line to add the "flathub" remote:
`  flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo`

flub replied:
Good idea, will do

flub commented:
On Tue 05 Mar 2019 at 12:15 -0800, muelli wrote:

>> Hey, I'm not really a fan of how you've created some commits in my name which I never made
>
> oh, I'm sorry about that. I didn't mean to offend. At all. In fact,
> quite the opposite. Should I force push a new history?

No offence taken, also no need to go change the history - certainly in
the main repo.


>> Lastly, as I indicated on the updated commit message, I currently think it's the flatpak's responsibility to provide the desktop integration files and not the deltachat-dekstop
>
> ah, this is unusual, because there are probably other downstreams
> which care about the `.desktop` file, the `.appdata` file, the icons,
> etc., too. AFAIU flatpaks try to be as lean as possible, shipping as
> few custom patches or files as possible. And spontaneously, I'd say
> that the upstream of a desktop app is the very correct place to handle
> those.  And it currently does, right?  If you build the `.deb` file
> then you get the icon. I would expect a .`desktop` file to be
> generated, too. It doesn't seem smart to make every downstream ship
> the same file themselves.

Oh, I see your point.  I was concentrating too much on getting rid of
the electron-builder and thus the upstream AppImage and .deb
generation.  But I guess it makes sense that eventually someone wants to
build a .deb or .rpm for a distribution and than having these files in
this repo instead of deltachat-desktop would be a pain.  I'll create a
PR or issue there to supply them without needing electron-builder and
update this side to mention they're temporary and may change.  I'm
thinking to still explicitly install them here though, as I would prefer
deltachat-desktop to not worry about installing things.  Does that sound
right to you?

I'll update this PR tomorrow probably.

Cheers

muelli commented:
>  I'm thinking to still explicitly install them here though, as I would prefer deltachat-desktop to not worry about installing things. Does that sound right to you?

I guess it depends on what the goals are. I'd say that a nice desktop app does indeed have a "make install" like command which takes care of the relevant files being pushed into their locations. That's at least how "all the existing apps" work. More or less. Look at [this](https://github.com/flathub/org.wireshark.Wireshark/blob/53dd239248af92a648ed1ed2dec5f965598dd0ca/org.wireshark.Wireshark.yml#L141), for example. Or, [here](https://github.com/flathub/org.gnome.Keysign/blob/749bd1b9056381e3edd5b929cfef923d5733730c/org.gnome.Keysign.yml#L269) for another app that I maintain. They don't require downstreams to manually move files around.

I understand upstreams that want to get rid of the responsibility of installing files, if only to reduce to maintenance burden. But in fairness, the files and their locations are quite standardised, so all downstreams usually take the same file and the same path which then makes it economically more useful to have the upstream package provide that `make install` command.

flub commented:
I've updated the readme, I think this is fine to merge again now.

As for the desktop files, in the ideal scenario you're right, but deltachat is so far away from this that we just want to make baby steps in the right direction.  I'd like to let deltachat-desktop only worry about having to create something which can run the application and not have to worry about how to correctly install for all the different platforms.  At least for the foreseeable future.  If other packagers come along then we'll work with them to remove duplication where it makes sense.

muelli commented:
> As for the desktop files, in the ideal scenario you're right, but deltachat is so far away from this

fair enough. 

Can you leave the sources in the bottom? That's how most of the manifests that I have encountered are written.

flub commented:
It's like top-posting for build stuff.^W^W^W^W^W^W^W  More seriously, is there some logic to this I'm missing?

Anyway, done.

muelli commented:
> is there some logic to this I'm missing?

dunno. Maybe it's due to the app's sources changing rather frequently and scrolling all the way to the bottom being more convenient.

I've re-added comments reminding me of the appdata file and the generated-sources.  Let's hope that the `--recursive` flag is indeed not needed any more. Otherwise I'll be spending weekends hunting weird failures down.
Thanks for the patches!

flub commented:
On Sun 10 Mar 2019 at 05:00 -0700, muelli wrote:

>> is there some logic to this I'm missing?
>
> dunno.

So why keep it this way?

> Maybe it's due to the app's sources changing rather frequently and scrolling all the way to the bottom being more convenient.

If you change the sources the build commands and everything else need to be reviewed too.

> Let's hope that the `--recursive` flag is indeed not needed any more. Otherwise I'll be spending weekends hunting weird failures down.

It's only a readme, you can keep the existing workflow with the makefile
you have.

> Thanks for the patches!

Thanks for helping things get into shape!

muelli commented:
> So why keep it this way?

For my part, because it makes dealing with the files easier. I have a few apps to maintain and if everything is organised the same, it's less mental effort for me. And I believe that others can take over more easily if they see a structure that they are used to.






> If you change the sources the build commands and everything else need to be reviewed too.

most often not, though. The velocity of build system changes is not so high. At least for traditional apps. I reckon the workflow for most people is to update the sources, push, and wait for the bot to complain.

muelli commented:
FTR: https://github.com/electron-userland/electron-builder/issues/1993 is probably relevant for the app if it wants to include appdata in the build.

flub commented:
I created https://github.com/deltachat/deltachat-desktop/issues/729 for it.
